### DISCUSSION TITLE
I'd like to ask you about cwe_checker performance.

### DISCUSSION THREAD

**Discussion by mksonzien** [fkie-cad/cwe_checker#260] [Q&A] [answered]
I'd like to improve the speed of the cwe-checker, but I'd like to ask if there's a way.
It is easy to analyze binary, so we will improve it.
Currently, it takes 40 seconds based on 200 KB, and I want to reduce the time as much as possible.
Is there any other idea or way to use a particular option?
(It may not be natural because the native language is not English)

**Comment by Enkelmann** [comment, marked as answer]
If you are asking whether one can speed up analysis time by a smart choice of analysis parameters, then it is not really possible to speed up the analysis at the moment. At least I am not aware of any parameters that one could change to speed up the computation without sacrificing analysis quality for it. You can restrict the computed analyses via the `--partial` command line option to only those that you really need, but since the bulk of the analysis time is spent in the `PointerInference` analysis (which is the base for many other analyses), this will not save a lot of time in general.

If you are asking how one could improve the analysis code to speed up computation, then there are some ways. At the time of this comment there are two main areas for possible runtime improvements:
* The performance of the `PointerInference` analysis. Right now the fixpoint that is computed by the analysis has some convergence problems, especially for the version in the master branch. I intend to rework the analysis soon to fix the convergence issues. Since one needs some intricate knowledge about how the analysis works, I fear that the necessary refactoring would be difficult to do for outside contributors at this time.
* Right now all fixpoint computations are done without any parallelization (and thus using only a single CPU core). So an obvious way to improve the speed is to parallelize fixpoint computations. If one wants to do this in a sound and deterministic way then the problem is also not easy (people have written research papers about the topic) but doable with only minimal knowledge about the cwe_checker internals. I think one could also implement a simple non-deterministic parallelization scheme without too much effort for cases, where determinism is not that important. If someone wants to tackle this I would be happy to help and mentor the effort!

**Reply by mksonzien** [reply]
Thank you.
I solved the problem by running the binary in parallel.

**Reply by Enkelmann** [reply]
A little advice for other people running multiple instances of the cwe_checker in parallel: Keep an eye on the RAM usage! For larger binaries a single instance of the cwe_checker may use several GB of RAM. So one should restrict the maximum number of parallel instances of the cwe_checker according to the amount of RAM available on the system.

**Reply by mksonzien** [reply]
Thank you for your advice :)

**Reply by ZeroXLN** [reply]
> Thank you. I solved the problem by running the binary in parallel.

Hello, can you share the details on parallelization？
